### PR TITLE
feat: link to docs in error msgs via docLink template string

### PR DESCRIPTION
**Pull request by marcoscaceres**
The `docLink` tagged template string allows one to more easily link to Respec documentation in error and warnings by automatically linking up keywords.

```JS
docLink`For example, the ${"[specStatus]"}  is required.`
docLink`For example, the ${"[specStatus|#custom-id]"}  is required.`
```

### CODE REVIEW

**Comment by sidvishnoi** [comment]
I like how clever it is, and it also feels like a good use of tagged templates, but I think we can use something simpler here. Particularly, I don't like hard-coding all the config options and other links in one place - it'll be maintenance burden, can get outdated, and also unnecessarily increases bundle size.

May I suggest simpler approaches?:
```js
const current = `Please use the ${docLink("group")} configuration option instead.`;
const thisPR = docLink`Please use the ${"group"} configuration option instead.`;

const alt1 =  `Please use the ${docLink("group", "optionalCustomIdOrURL")} configuration option instead.`;
const alt2 = docLink`Please use the ${["group", "optionalCustomId"]} configuration option instead.`; 
const alt3 = docLink`Please use the ${"group|optionalCustomId"} configuration option instead.`;
```

**Comment by marcoscaceres** [comment]
We discussed. Going with 3.

File: src/core/respec-docs.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+]);
+
+/**
+ * Tagged Template string, helps with linking to documentation.
+ */
+export function docLink(strings, ...keys) {
+  return strings
+    .map((s, i) => {
+      const key = keys[i] ?? "";
+      switch (true) {
```
Comment (sidvishnoi) [inline src/core/respec-docs.js:83]: Pattern matching for us peasants 😁

File: src/core/caniuse.js [hunk 2]
```diff
@@ -104,10 +104,8 @@ function getNormalizedConf(conf) {
   if (Array.isArray(browsers)) {
     const invalidBrowsers = browsers.filter(browser => !BROWSERS.has(browser));
     if (invalidBrowsers.length) {
-      const names = invalidBrowsers.map(b => `"\`${b}\`"`).join(", ");
-      const msg =
-        `Ignoring invalid browser(s): ${names} in ` +
-        "[`respecConfig.caniuse.browsers`](https://github.com/w3c/respec/wiki/caniuse)";
+      const names = joinAnd(invalidBrowsers.map(b => `"\`${b}\`"`));
```
Comment (sidvishnoi) [inline src/core/caniuse.js:107]: For consistently using `mapper` arg..
```suggestion
      const names = joinAnd(invalidBrowsers, b => `"\`${b}\`"`);
```

Comment (sidvishnoi) [inline src/core/caniuse.js:107]: For consistently using `mapper` arg..
```suggestion
      const names = joinAnd(invalidBrowsers, b => `"\`${b}\`"`);
```

File: src/core/data-tests.js [hunk 3]
```diff
@@ -139,8 +133,8 @@ function handleDuplicates(testURLs, elem) {
     (link, i, self) => self.indexOf(link) !== i
   );
   if (duplicates.length) {
-    const msg = `Duplicate tests found`;
-    const hint = `To fix, remove duplicates from "data-tests": ${duplicates
+    const msg = `Duplicate tests found in the ${"data-tests"} attribute`;
```
Comment (sidvishnoi) [inline src/core/data-tests.js:136]:
```suggestion
    const msg = docLink`Duplicate tests found in the ${"data-tests"} attribute`;
```

File: src/core/caniuse.js (hunk 2 above)
Comment (sidvishnoi) [reply src/core/caniuse.js:107]: Update: Can use `mdJoinAnd` instead.

File: src/core/respec-docs.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+export function docLink(strings, ...keys) {
+  return strings
+    .map((s, i) => {
+      const key = keys[i];
+      if (!key) {
+        return s;
+      }
+      const [linkingText, href] = key.split("|");
+      // Known things...
+      if (!href && linkableThings.has(linkingText)) {
```
Comment (sidvishnoi) [inline src/core/respec-docs.js:100]: Not particularly happy with this idea 😞

**Review by sidvishnoi — COMMENTED** [review]
Not particularly happy with the idea of listing all known config options in `core/respec-docs` 😞 
But liking the general idea.

File: src/core/caniuse.js (hunk 2 above)
Comment (marcoscaceres) [reply src/core/caniuse.js:107]: As a followup, I need to create a new "quoted code mapper". I'll add the suggestion for now tho.

File: src/core/respec-docs.js (hunk 4 above)
Comment (marcoscaceres) [reply src/core/respec-docs.js:100]: how about `${"[linkable]"}` or `${"!linkable"}`? Then we don't need the check. We just need a way to distinguish them from regular text.

Comment (sidvishnoi) [reply src/core/respec-docs.js:100]: Yeah! Either of above sounds good. "[linkable]" looks similar to a markdown link, while "!linkable" is shorter, so either is good.

Comment (marcoscaceres) [reply src/core/respec-docs.js:100]: Ok, let's go with `[...]`, just for familiarity.

File: src/core/linter-rules/no-http-props.js [hunk 5]
```diff
@@ -3,18 +3,18 @@
...
-import { getIntlData, joinAnd, showWarning } from "../utils.js";
+import { docLink, getIntlData, joinAnd, showWarning } from "../utils.js";
 
 const ruleName = "no-http-props";
 export const name = "core/linter-rules/no-http-props";
 
 const localizationStrings = {
   en: {
-    msg: "Insecure URLs are not allowed in `respecConfig`.",
+    msg: docLink`Insecure URLs are not allowed in ${"[`respecConfig`|#configuring-respec]"}.`,
```
Comment (sidvishnoi) [inline src/core/linter-rules/no-http-props.js:13]: Maybe we can add `respecConfig` as custom ID in docs instead?

File: src/core/linter-rules/no-http-props.js [hunk 6]
```diff
@@ -40,7 +40,7 @@ export function run(conf) {
     );
 
   if (offendingMembers.length) {
-    const keys = joinAnd(offendingMembers, key => `\`${key}\``);
+    const keys = joinAnd(offendingMembers, key => docLink`${`[${key}]`}`);
```
Comment (sidvishnoi) [inline src/core/linter-rules/no-http-props.js:43]: Too many ` 😄 

```suggestion
    const keys = joinAnd(offendingMembers, key => docLink`${toMDCode([${key}])}`);
```

File: src/core/utils.js [hunk 7]
```diff
@@ -900,10 +900,63 @@ export function showWarning(message, pluginName, options = {}) {
...
+ * Makes a string `coded`.
  *
- * @param {string} prop ReSpec configuration property to link to in docs.
+ * @param {string} item
+ * @returns {string}
  */
-export function docLink(prop) {
-  return `[\`${prop}\`](https://respec.org/docs/#${prop})`;
+export function toMDCode(item) {
+  return String(item) ? `\`${item}\`` : "";
```
Comment (sidvishnoi) [inline src/core/utils.js:909]: JSDoc + TS will ensure we only pass a string
```suggestion
  return item ? `\`${item}\`` : "";
```

File: src/core/utils.js [hunk 8]
```diff
@@ -900,10 +900,63 @@ export function showWarning(message, pluginName, options = {}) {
...
+}
+
+/**
+ * Wraps in back-ticks ` for code.
+ *
+ * @param {string[]} array
+ * @param {object} options
+ * @param {boolean} options.quotes Surround each item in quotes
+ */
+export function mdJoinOr(array, { quotes } = { quotes: false }) {
```
Comment (sidvishnoi) [inline src/core/utils.js:919]: Maybe `codedJoinOr`?

File: src/core/utils.js [hunk 9]
```diff
@@ -900,10 +900,63 @@ export function showWarning(message, pluginName, options = {}) {
...
+}
+
+function addQuotes(item) {
+  return String(item) ? `"${item}"` : "";
+}
+
+/**
+ * Tagged template string, helps with linking to documentation.
+ * Things inside [squareBrackets] are considered direct links to the documentation.
+ * To alias something, once can use a "|", like [respecConfig|#respec-configuration].
```
Comment (sidvishnoi) [inline src/core/utils.js:941]:
```suggestion
 * To alias something, one can use a "|", like [respecConfig|#respec-configuration].
```

File: src/w3c/headers.js [hunk 10]
```diff
@@ -569,14 +559,12 @@ export function run(conf) {
...
-    showError(msg, name);
+    const msg = docLink`If one of ${"[wg]"}, ${"[wgURI]"}, or ${"[wgPatentURI]"} is an array, they all have to be.`;
+    const hint = docLink`Use the ${"[group]"} option with an array instead.`;
+    showError(msg, name, { hint });
   }
   if (conf.isCGBG && !conf.wg) {
-    const msg =
-      "[`wg`](https://github.com/w3c/respec/wiki/wg)" +
-      " configuration option is required for this kind of document.";
+    const msg = docLink`][wg]"} configuration option is required for this kind of document.`;
```
Comment (sidvishnoi) [inline src/w3c/headers.js:567]:
```suggestion
    const msg = docLink`${"[wg]"} configuration option is required for this kind of document.`;
```

File: tests/spec/core/utils-spec.js [hunk 11]
```diff
@@ -612,5 +612,71 @@ describe("Core - Utils", () => {
...
+
+    describe("toMDCode()", () => {
+      it("wraps a string in backticks", () => {
+        expect(utils.toMDCode("")).toBe("");
+        expect(utils.toMDCode("test")).toBe("`test`");
+      });
+    });
+
+    describe("mdJoinOr()", () => {
+      it("uses uses disjunction", () => {
```
Comment (sidvishnoi) [inline tests/spec/core/utils-spec.js:624]:
```suggestion
      it("uses disjunction", () => {
```

File: tests/spec/core/utils-spec.js [hunk 12]
```diff
@@ -612,5 +612,71 @@ describe("Core - Utils", () => {
...
+        expect(utils.mdJoinAnd(["a", "b", "c"], { quotes: true })).toBe(
+          '`"a"`, `"b"`, and `"c"`'
+        );
+      });
+    });
+    describe("docLink", () => {
+      const docLink = utils.docLink;
+      it("it allows unlinked strings", () => {
+        const result = docLink`Link to ${"nothing"}.`;
+        expect(result).toBe("Link to nothing.");
```
Comment (sidvishnoi) [inline tests/spec/core/utils-spec.js:652]:
```suggestion
        const result = docLink`Link to ${"nothing"} ${1+2}.`;
        expect(result).toBe("Link to nothing 3.");
```

File: tests/spec/core/utils-spec.js [hunk 13]
```diff
@@ -612,5 +612,71 @@ describe("Core - Utils", () => {
...
+      });
+
+      it("it links known config options", () => {
```
Comment (sidvishnoi) [inline tests/spec/core/utils-spec.js:655]:
```suggestion
      it("it links to [config] options", () => {
```

File: src/core/utils.js [hunk 14]
```diff
@@ -900,10 +900,63 @@ export function showWarning(message, pluginName, options = {}) {
...
+ * @returns {string}
  */
-export function docLink(prop) {
-  return `[\`${prop}\`](https://respec.org/docs/#${prop})`;
+export function toMDCode(item) {
+  return String(item) ? `\`${item}\`` : "";
+}
+
+/**
+ * Wraps in back-ticks ` for code.
```
Comment (sidvishnoi) [inline src/core/utils.js:913]:
```suggestion
 * Joins an array of strings, wrapping each string in back-ticks (`) for inline markdown code.
```

File: src/core/utils.js [hunk 15]
```diff
@@ -900,10 +900,63 @@ export function showWarning(message, pluginName, options = {}) {
...
+
+function addQuotes(item) {
+  return String(item) ? `"${item}"` : "";
+}
+
+/**
+ * Tagged template string, helps with linking to documentation.
+ * Things inside [squareBrackets] are considered direct links to the documentation.
+ * To alias something, once can use a "|", like [respecConfig|#respec-configuration].
+ *
```
Comment (sidvishnoi) [inline src/core/utils.js:942]:
```suggestion
 * @param {TemplateStringsArray} strings
 * @param {string[]} keys
```

File: src/core/utils.js [hunk 16]
```diff
@@ -900,10 +900,63 @@ export function showWarning(message, pluginName, options = {}) {
...
+ */
+export function docLink(strings, ...keys) {
+  return strings
+    .map((s, i) => {
+      const key = keys[i];
+      if (!key) {
+        return s;
+      }
+      // Linkables are wrapped in square brackets
+      if (!key.startsWith("[") && !key.endsWith("]")) return s + key;
```
Comment (sidvishnoi) [inline src/core/utils.js:952]:
```suggestion
      if (!key.startsWith("[") && !key.endsWith("]")) {
        return s + key;
      }
```

File: tests/spec/core/utils-spec.js (hunk 12 above)
Comment (sidvishnoi) [reply tests/spec/core/utils-spec.js:652]: We probably don't (want to) support this?

**Review by sidvishnoi — APPROVED** [review]
Almost there 😃

File: tests/spec/core/utils-spec.js (hunk 12 above)
Comment (marcoscaceres) [reply tests/spec/core/utils-spec.js:652]: This might throw, so I'll avoid this for now.

File: src/core/linter-rules/no-http-props.js (hunk 5 above)
Comment (marcoscaceres) [reply src/core/linter-rules/no-http-props.js:13]: I don't mind. Can certainly do later.

File: src/core/utils.js (hunk 8 above)
Comment (marcoscaceres) [reply src/core/utils.js:919]: yeah that might be nicer.

File: src/core/linter-rules/no-http-props.js (hunk 5 above)
Comment (sidvishnoi) [reply src/core/linter-rules/no-http-props.js:13]: I edited wiki. Now we've https://respec.org/docs/#respecConfig

Comment (sidvishnoi) [reply src/core/linter-rules/no-http-props.js:13]: (we've both actually)

File: src/w3c/group.js [hunk 17]
```diff
@@ -23,18 +24,18 @@ export async function run(conf) {
...
     return;
   }
 
   if (usedLegacyOptions.length) {
-    const outdatedOptionsStr = joinAnd(usedLegacyOptions, s => `\`${s}\``);
-    const msg = `Configuration options ${outdatedOptionsStr} are superseded by \`group\` and will be overridden by ReSpec.`;
-    const hint = "Please remove them from `respecConfig`.";
+    const outdatedOptionsStr = codedJoinAnd(usedLegacyOptions);
+    const msg = docLink`Configuration options ${outdatedOptionsStr} are superseded by ${"[group]"} and will be overridden by ReSpec.`;
+    const hint = docLink`Remove them from the document's ${"[respecConfig|#configuring-respec]"} to silence this warning.`;
```
Comment (sidvishnoi) [inline src/w3c/group.js:38]:
```suggestion
    const hint = docLink`Remove them from the document's ${"[respecConfig]"} to silence this warning.`;
```

**Review by sidvishnoi — APPROVED** [review]
:shipit:

**Pull request merged**